addon/utils/html-input-parser.ts

```diff
@@ -0,0 +1,155 @@
+    this.safeTags = safeTags ? safeTags : DEFAULT_SAFE_TAGS;
+    this.uriSafeAttributes = uriSafeAttributes ? uriSafeAttributes : DEFAULT_URI_SAFE_ATTRIBUTES;
+    this.tagMap = tagMap ? tagMap : DEFAULT_TAG_MAP;
+  }
+
+  /**
+   * Takes an html string, preprocesses its nodes and sanitizes the result.
+   * Returns the cleaned html string.
+   *
+   * @method cleanupHTML
+   */
+  cleanupHTML(htmlString: string): string {
+    const parser = new DOMParser();
+    const document = parser.parseFromString(htmlString, "text/html");
+    const rootNode = document.body;
+    rootNode.normalize();
+
+    const preprocessedNode = this.preprocessNode(rootNode) as HTMLElement; // Assumption: root node can never be text
+    return DOMPurify.sanitize(preprocessedNode.innerHTML, {
+      ALLOWED_TAGS: this.safeTags,
+      ALLOWED_ATTR: this.safeAttributes,
+      // ADD_URI_SAFE_ATTR: this.uriSafeAttributes TODO: does this work?
+    });
+  }
+
+  /**
+   * Preprocess all nodes replacing the tag if it appears on the tagMap variable
+   * and adds the lumpNode property if needed.
+   * TODO: check if we can get rid of casts
+   *
+   * @method preprocessNode
+   */
+  preprocessNode(node: Node) {
+    let cleanedNode = node.cloneNode();
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const tag = tagName(node);
+      const tagMapping = this.tagMap.get(tag);
+
+      // If we have to replace the tag name we create another node with the new
+      // tag name and copy all the attribute of the original node.
+      if (tagMapping) {
+        cleanedNode = document.createElement(tagMapping);
+        this.copyAllAttributes((node as HTMLElement), (cleanedNode as HTMLElement));
+      } else if (tag === "a" && !(cleanedNode as HTMLLinkElement).href) {
+        cleanedNode = document.createElement("span");
+        this.copyAllAttributes((node as HTMLElement), (cleanedNode as HTMLElement));
+      }
+
+      // Clean all children of node.
+      cleanedNode.textContent = "";
+      if (this.lumpTags.includes(tag)) {
+        (cleanedNode as HTMLElement).setAttribute(
+          "property",
+          "http://lblod.data.gift/vocabularies/editor/isLumpNode"
+        );
+      }
+
+      if (node.hasChildNodes()) {
+        for (let i = 0; i < node.childNodes.length; i++) {
+          const cleanedChild = this.preprocessNode(node.childNodes[i]);
+
+          if (cleanedChild) {
+            if (this.lumpTags.includes(tag)) {
+              // Make sure we can place the cursor before the non editable element.
+              cleanedNode.appendChild(document.createTextNode(""));
+            }
+
+            cleanedNode.appendChild(cleanedChild);
+
+            if (this.lumpTags.includes(tag)) {
+              // Make sure we can place the cursor after the non editable element.
+              cleanedNode.appendChild(document.createTextNode(""));
+            }
+          }
+        }
+      }
+    } else if (node.nodeType === Node.TEXT_NODE) {
```

Comment:
there should be a similar helper to isElement for textnodes. If there isn't, make one.